### PR TITLE
SPARK-4644 blockjoin

### PR DESCRIPTION
**Pull request by koertkuipers**
Although the discussion (and design doc) under SPARK-4644 seem focussed on other aspects of skew (OOM mostly) than this pullreq (which focusses on avoiding a single reducer taking a long time), i decided to put this pullreq under SPARK-4644 anyhow, to avoid the proliferation of JIRA tickets. If this is not the right place let me know and i will move it.

Inspired by block join in scalding.
From scalding docs:

This is useful in cases where the data has extreme skew. A symptom of this is that we may see a job stuck for a very long time on a small number of reducers.
A block join is way to get around this: we add a random integer field and a replica field to every tuple in the left and right pipes. We then join on the original keys and on these new dummy fields. These dummy fields make it less likely that the skewed keys will be hashed to the same reducer.

The final data size is right \* rightReplication + left \* leftReplication but because of the fragmentation, we are guaranteed the same number of hits as the original join.

### CODE REVIEW

**Comment by andrewor14** [comment]
add to whitelist

File: core/src/main/scala/org/apache/spark/api/java/JavaPairRDD.scala [hunk 1]
```diff
@@ -515,6 +515,76 @@ class JavaPairRDD[K, V](val rdd: RDD[(K, V)])
   }
 
   /**
+   * Same as join, but uses a block join, otherwise known as a replicate fragment join.
+   * This is useful in cases where the data has extreme skew.
+   * The input params leftReplication and rightReplication control the replication of the left
+   * (this rdd) and right (other rdd) respectively.
+   */
+  def blockJoin[W](other: JavaPairRDD[K, W], leftReplication: Int, rightReplication: Int, 
+    partitioner: Partitioner): JavaPairRDD[K, (V, W)] = {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/api/java/JavaPairRDD.scala]: style:
```
def blockJoin[W](
    other: JavaPairRDD[K, W],
    leftReplication: Int,
    ...): JavaPairRDD[...] = {
  ...
}
```

File: core/src/main/scala/org/apache/spark/api/java/JavaPairRDD.scala [hunk 2]
```diff
@@ -515,6 +515,76 @@ class JavaPairRDD[K, V](val rdd: RDD[(K, V)])
...
+  /**
+   * Same as leftOuterJoin, but uses a block join, otherwise known as a replicate fragment join.
+   * This is useful in cases where the data has extreme skew.
+   * The input param rightReplication controls the replication of the right (other rdd).
+   */
+  def blockLeftOuterJoin[W](other: JavaPairRDD[K, W], rightReplication: Int,
+    partitioner: Partitioner): JavaPairRDD[K, (V, Optional[W])] = {
+    fromRDD(rdd.blockLeftOuterJoin(other, rightReplication, partitioner).mapValues{ case (v, w) =>
+      (v, JavaUtils.optionToOptional(w))
+    })
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/api/java/JavaPairRDD.scala]: need space after `.mapValues`

File: core/src/main/scala/org/apache/spark/api/java/JavaPairRDD.scala [hunk 3]
```diff
@@ -515,6 +515,76 @@ class JavaPairRDD[K, V](val rdd: RDD[(K, V)])
...
+  }
+
+  /**
+   * Same as leftOuterJoin, but uses a block join, otherwise known as a replicate fragment join.
+   * This is useful in cases where the data has extreme skew.
+   * The input param rightReplication controls the replication of the right (other rdd).
+   */
+  def blockLeftOuterJoin[W](other: JavaPairRDD[K, W], rightReplication: Int,
+    partitioner: Partitioner): JavaPairRDD[K, (V, Optional[W])] = {
+    fromRDD(rdd.blockLeftOuterJoin(other, rightReplication, partitioner).mapValues{ case (v, w) =>
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/api/java/JavaPairRDD.scala]: need space after `.mapValues`

**Comment by andrewor14** [comment]
@zsxwing

**Comment by zsxwing** [comment]
@rxin I remember you said you would like such improvement to be added to Spark SQL rather than Spark Core. What's your thoughts on this one?

**Comment by koertkuipers** [comment]
i put this in a spark package together with skewjoin in case anyone wants to use it.
see here:
http://spark-packages.org/package/tresata/spark-skewjoin
https://github.com/tresata/spark-skewjoin

**Comment by JoshRosen** [comment]
Given that we're unlikely to add this in core right now and given that it's available in a Spark Package, would you mind closing this PR for now in order to clean up the review backlog? Thanks!

**Comment by koertkuipers** [comment]
sure

**Pull request closed without merging**